### ISSUE TITLE
Sentry not grouping exceptions with identical stack trace

### ISSUE THREAD

**Issue by jiri-matejka** [getsentry/sentry#5705]
We are using Sentry to catch exceptions from C# desktop application. Sentry is not grouping exceptions of type  with identical stack trace and identical type (System.NotImplementedException). OS differs.
The messages are different because of localization of this standard message:
"La méthode ou l'opération n'est pas implémentée." vs "The method or operation is not implemented."
But according to [this comment](https://github.com/getsentry/sentry/issues/4710#issuecomment-269421856), issues are not grouped by the message.

Is this a bug or am I missing other properties that affect grouping?

Stack trace:
`MailClient.Accounts.AccountManager in MailClient.Accounts.ISchedulingAccount.get_SentByAddress
MailClient.UI.Forms.formEvent.updateOrganizer+5e
MailClient.UI.Forms.formEvent.controlAttendeeTextBox_AttendeesAdded(System.Object sender, MailClient.UI.Controls.ControlAttendeeTextBox+AttendeesAddedEventArgs e)+4b
MailClient.UI.Controls.ControlAttendeeTextBox.OnAttendeesAdded(MailClient.UI.Controls.ControlAttendeeTextBox+AttendeesAddedEventArgs e)+8
MailClient.UI.Controls.ControlAttendeeTextBox.addAttendeeFromContactWithAddress(MailClient.UI.Controls.ControlContactTextBox+ContactWithAddress contactWithAddr)+46
MailClient.UI.Controls.ControlAttendeeTextBox.AcceptSuggestion(MailClient.UI.Controls.ControlContactTextBox+ContactSuggestion item)
MailClient.UI.Controls.ControlContactTextBox.ChooseSuggestion(MailClient.UI.Controls.ControlContactTextBox+ContactSuggestion item, MailClient.UI.Controls.ConfirmationSource confirmationSource)+16c
MailClient.UI.Controls.ControlSuggestionTextBox``1.chooseSuggestion(MailClient.UI.Controls.ConfirmationSource confirmationSource)+25
MailClient.UI.Controls.ControlSuggestionTextBox``1.dataGrid_CellClicked(System.Object sender, MailClient.UI.Controls.ControlDataGrid.DataGridCellClickedEventArgs e)+11 (frame repeated 1 times)
MailClient.UI.Controls.ControlDataGrid.ControlDataGrid.handleMouseClick(System.Windows.Forms.MouseEventArgs e, MailClient.UI.Controls.ControlDataGrid.ControlDataGrid+MouseClickSource source)+8d8
MailClient.UI.Controls.ControlDataGrid.ControlDataGrid.OnMouseDown(System.Windows.Forms.MouseEventArgs e)+a5
System.Windows.Forms.Control.WmMouseDown(System.Windows.Forms.Message& m, System.Windows.Forms.MouseButtons button, System.Int32 clicks)+95`

Sentry version 8.17.0, inside Docker.

**Comment by dcramer** [comment]
Can you dump the JSON for the stack? If its truly identical and the classname of the exception is identical (or not present) it will aggregate together.

**Comment by jiri-matejka** [comment]
Ok, how can I do it?

**Comment by dcramer** [comment]
Click the JSON link in the UI

**Comment by jiri-matejka** [comment]
Ok, as it turns out they really have different stack trace differing in the instruction pointer, see below in bold. 
I have many more such events but I have unfortunately merged them together already so I can't tell now whether they were previously grouped by Sentry.

Two questions:
1. Is there a way to ignore instruction offset differences in these low level procedures while grouping? (WndProc is Windows event loop handler).
2. Any idea why one has as instruction_offset property and the other has not? They are reported in the absolutely same way.

Thank your for your help.

{
	"context_line": "System.Windows.Forms.ScrollableControl.WndProc(System.Windows.Forms.Message& m)**+34**",
	"function": "WndProc",
	"in_app": false,
	"lineno": 0,
	"module": "System.Windows.Forms.ScrollableControl"
},
{
	"context_line": "System.Windows.Forms.Control.WndProc(System.Windows.Forms.Message& m)**+66f**",
	"function": "WndProc",
	"in_app": false,
	"lineno": 0,
	"module": "System.Windows.Forms.Control"
}

vs.

{
	"context_line": "System.Windows.Forms.ScrollableControl.WndProc(System.Windows.Forms.Message& m)**+3a**",
	"function": "WndProc",
	"in_app": false,
	**"instruction_offset": 58**,
	"lineno": 0,
	"module": "System.Windows.Forms.ScrollableControl"
},
{
	"context_line": "System.Windows.Forms.Control.WndProc(System.Windows.Forms.Message& m)**+5fe**",
	"function": "WndProc",
	"in_app": false,
	**"instruction_offset": 1534**,
	"lineno": 0,
	"module": "System.Windows.Forms.Control"
}

**Comment by dcramer** [comment]
What we should do is parse it out on the server (as I'm assuming it is changing in the context_line of each one). Alternatively the client could do it as well.

**Comment by filipnavara** [comment]
Ok, if the context_line is used then it could possibly explain why the exceptions don't group properly.

I checked the code for the hashing and there's something weird there though (src/sentry/interfaces/stacktrace.py):
`
  | elif self.function and self.is_unhashable_function():
  | can_use_context = True
  | else:
  | can_use_context = True
`

Not sure what was the intention of the code, but it doesn't make sense to check for a condition and then set can_use_context = True in both the "if" case and the "else" case.

**Comment by chrisharrisonkiwi** [comment]
I have noticed Sentry really struggles to group similar errors.
It's kind of a dealbreaker for the company I work for.

If the errors have to be 100% identical without any smart under the hood decision making going on we are going to have to jump ship to a competitor that rhymes with: "LayFun".

They have excellent grouping technology.

**Comment by dcramer** [comment]
tl;dr "i will use a closed source competitor because i think they're better" is not helpful

If you want to talk about facts:

Our heuristics go through hundreds and hundreds of branches to group issue together. I'd challenge everyone else's tech is modeled off of ours, especially given that Sentry's source code is completely public, and the analysis and grouping is not patented.

That said, if you're seeing specific areas where grouping fails you should open an issue about the specific case. We never require exact stacktraces, and even where we rely on stacktraces its much more complex than an identity check. It doesn't mean we handle every edge case well, but we do handle a LOT of edge cases. 

There is ongoing work to improve this even more, but generally the areas where we are weak are those of newer runtime support (like mobile or native apps). If you're seeing this elsewhere we need to actually know the cases due to the complexity of the algorithm.

For example, if you have two issues on sentry.io you send us the links and we can identify if they actually should be grouped together, and if they weren't, why not.

Otherwise I'd actually encourage you to try out other products, and areas where they're actually stronger than Sentry we'd love to know about.

**Comment by BYK** [comment]
Closing this issue due to staleness. Feel free to comment here if you think we should still work on this.
